### PR TITLE
Extract pubkey crate

### PR DESCRIPTION
**Pull request by kevinheavey**
#### Problem
`solana_program::pubkey` can get its own crate once #2015 is merged

#### Summary of Changes
- move pubkey.rs to its own crate and re-export in solana-program
- move relevant syscall definitions to the new crate and re-export them in their old location
- make the code requiring curve25519-dalek and sha2 optional (these are methods so they can't be moved to separate crates)
- make bytemuck and serde optional
- make std optional but enable it by default

### CODE REVIEW

File: sdk/hash/src/lib.rs [hunk 1]
```diff
@@ -202,7 +202,7 @@ impl Hash {
     }
 }
 
-#[cfg(test)]
+#[cfg(all(test, feature = "frozen-abi", RUSTC_WITH_SPECIALIZATION))]
```
Comment (joncinque) [inline sdk/hash/src/lib.rs]: Was this change intended? These tests should always run, and not just for frozen-abi builds

File: sdk/program/src/message/legacy.rs [hunk 2]
```diff
@@ -63,8 +63,8 @@ mod builtins {
         // the value is "false"), or might be in one of these lists (if the value is "true")
         pub static ref MAYBE_BUILTIN_KEY_OR_SYSVAR: [bool; 256] = {
             let mut temp_table: [bool; 256] = [false; 256];
-            BUILTIN_PROGRAMS_KEYS.iter().for_each(|key| temp_table[key.0[0] as usize] = true);
-            sysvar::ALL_IDS.iter().for_each(|key| temp_table[key.0[0] as usize] = true);
+            BUILTIN_PROGRAMS_KEYS.iter().for_each(|key| temp_table[key.to_bytes()[0] as usize] = true);
+            sysvar::ALL_IDS.iter().for_each(|key| temp_table[key.to_bytes()[0] as usize] = true);
```
Comment (joncinque) [inline sdk/program/src/message/legacy.rs]: `to_bytes()` creates a new array, whereas `as_ref()` takes it in-place, so `as_ref()` should be preferred wherever possible
```suggestion
            BUILTIN_PROGRAMS_KEYS.iter().for_each(|key| temp_table[key.as_ref()[0] as usize] = true);
            sysvar::ALL_IDS.iter().for_each(|key| temp_table[key.as_ref()[0] as usize] = true);
```

File: sdk/program/src/message/legacy.rs [hunk 3]
```diff
@@ -76,7 +76,7 @@ mod builtins {
 )]
 #[allow(deprecated)]
 pub fn is_builtin_key_or_sysvar(key: &Pubkey) -> bool {
-    if MAYBE_BUILTIN_KEY_OR_SYSVAR[key.0[0] as usize] {
+    if MAYBE_BUILTIN_KEY_OR_SYSVAR[key.to_bytes()[0] as usize] {
```
Comment (joncinque) [inline sdk/program/src/message/legacy.rs]: same here: to_bytes() -> as_ref()
```suggestion
    if MAYBE_BUILTIN_KEY_OR_SYSVAR[key.as_ref()[0] as usize] {
```

**Comment by kevinheavey** [comment]
For checking the feature powerset I've used `cargo hack check --feature-powerset --exclude-features frozen-abi --all-targets` and  `cargo hack check --feature-powerset --exclude-features frozen-abi --all-targets --at-least-one-of 'std,'  --target wasm32-unknown-unknown`

File: sdk/pubkey/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+js-sys = { workspace = true }
+wasm-bindgen = { workspace = true }
+
+[dev-dependencies]
+anyhow = { workspace = true }
+arbitrary = { workspace = true, features = ["derive"] }
+bs58 = { workspace = true, features = ["alloc"] }
+rand = { workspace = true }
+# circular dev deps need to be path deps for `cargo publish` to be happy
+solana-program = { path = "../program" }
```
Comment (joncinque) [inline sdk/pubkey/Cargo.toml]: nit: Can you specifically call out that the circular dependency is for the doctests?

I'm not sure there's a better place for those doctests, so we might need to live with it unfortunately, but at least we'll know why it's there

File: sdk/pubkey/src/lib.rs [hunk 5]
```diff
@@ -1,25 +1,40 @@
...
+#![cfg_attr(RUSTC_WITH_SPECIALIZATION, feature(min_specialization))]
 #![allow(clippy::arithmetic_side_effects)]
 
-#[cfg(target_arch = "wasm32")]
-use crate::wasm_bindgen;
-#[cfg(test)]
+#[cfg(any(feature = "std", target_arch = "wasm32"))]
+extern crate std;
+#[cfg(any(test, feature = "dev-context-only-utils"))]
 use arbitrary::Arbitrary;
```
Comment (joncinque) [inline sdk/pubkey/src/lib.rs]: Since this is now gated with `dev-context-only-utils`, can we remove the `test` requirement? This seemed to work on my side
```suggestion
#[cfg(feature = "dev-context-only-utils")]
use arbitrary::Arbitrary;
```

File: sdk/pubkey/src/lib.rs [hunk 6]
```diff
@@ -30,18 +45,74 @@ pub const MAX_SEEDS: usize = 16;
...
 const MAX_BASE58_LEN: usize = 44;
 
+#[cfg(any(target_os = "solana", feature = "sha2", feature = "curve25519"))]
 const PDA_MARKER: &[u8; 21] = b"ProgramDerivedAddress";
 
-#[derive(Error, Debug, Serialize, Clone, PartialEq, Eq, FromPrimitive, ToPrimitive)]
+/// Copied from `solana_program::entrypoint::SUCCESS`
+/// to avoid a `solana_program` dependency
+#[cfg(target_os = "solana")]
+const SUCCESS: u64 = 0;
```
Comment (joncinque) [inline sdk/pubkey/src/lib.rs:54]: This is ok for this PR, but I think we should move this down into solana-define-syscall in future work

File: sdk/pubkey/src/lib.rs [hunk 7]
```diff
@@ -73,39 +144,71 @@ impl From<u64> for PubkeyError {
...
-    Pod,
-    Serialize,
-    Zeroable,
-)]
-#[cfg_attr(test, derive(Arbitrary))]
+#[cfg_attr(all(feature = "borsh", feature = "std"), derive(BorshSchema))]
+#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
+#[cfg_attr(feature = "bytemuck", derive(Pod, Zeroable))]
+#[derive(Clone, Copy, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
+#[cfg_attr(any(test, feature = "dev-context-only-utils"), derive(Arbitrary))]
```
Comment (joncinque) [inline sdk/pubkey/src/lib.rs]: Same here, let's remove the cfg test allowance
```suggestion
#[cfg_attr(feature = "dev-context-only-utils", derive(Arbitrary))]
```

File: sdk/pubkey/src/lib.rs [hunk 8]
```diff
@@ -30,18 +45,74 @@ pub const MAX_SEEDS: usize = 16;
...
+            Some(PubkeyError::IllegalOwner)
+        } else {
+            None
+        }
+    }
+    #[inline]
+    fn from_u64(n: u64) -> Option<Self> {
+        Self::from_i64(n as i64)
+    }
+}
```
Comment (joncinque) [inline sdk/pubkey/src/lib.rs:100]: Confirming that these match what's generated by num-derive: https://docs.rs/num-derive/latest/src/num_derive/lib.rs.html

For the `FromPrimitive` implementation, is there a way we can guarantee that the check is exhaustive? The best I can come up with is adding a new test to do the conversion.

File: sdk/pubkey/src/lib.rs [hunk 9]
```diff
@@ -30,18 +45,74 @@ pub const MAX_SEEDS: usize = 16;
...
+                f.write_str("Provided seeds do not result in a valid address")
+            }
+            PubkeyError::IllegalOwner => f.write_str("Provided owner is not allowed"),
+        }
+    }
+}
+
 impl<T> DecodeError<T> for PubkeyError {
     fn type_of() -> &'static str {
         "PubkeyError"
```
Comment (joncinque) [inline sdk/pubkey/src/lib.rs:121]: I can't comment further down, but the `From<u64>` implementation is missing `IllegalOwner` -- do you mind adding it?

File: sdk/pubkey/src/lib.rs [hunk 10]
```diff
@@ -73,39 +144,71 @@ impl From<u64> for PubkeyError {
...
+            Some(ParsePubkeyError::Invalid)
+        } else {
+            None
+        }
+    }
+    #[inline]
+    fn from_u64(n: u64) -> Option<Self> {
+        Self::from_i64(n as i64)
+    }
+}
```
Comment (joncinque) [inline sdk/pubkey/src/lib.rs:205]: Same as with the other `FromPrimitive` implementation, is there a way to add an exhaustiveness check? If not, would you mind adding a simple test to make sure that the conversion works? ie

```
        macro_rules! ensure_mapping {
            ($ty: path, $($num: literal => $variant: path),+ $(,)?) => {
                // assert that the given values produce the expected variants
                $(assert_eq!(<$ty>::from($num), $variant);)+
                $(assert_eq!(<$ty>::from_i64($num), Some($variant));)+
                $(assert_eq!(<$ty>::from_u64($num), Some($variant));)+

                // this generated fn will never be called but will produce a
                // non-exhaustive pattern error if you've missed a variant
                #[allow(dead_code)]
                fn check_all_covered(err: $ty) {
                    match err {
                        $($variant => {})+
                    };
                }
            }
        }
    #[test]
    fn test_exhaustive_conversion() {
        ensure_mapping! {
            PubkeyError,
            0 => PubkeyError::MaxSeedLengthExceeded,
            1 => PubkeyError::InvalidSeeds,
            2 => PubkeyError::IllegalOwner,
        }
        ensure_mapping! {
            ParsePubkeyError,
            0 => ParsePubkeyError::WrongSize,
            1 => ParsePubkeyError::Invalid,
        }
    }
```

You'll need to add a `From<u64>` implementation for `ParsePubkeyError` to get this to work, unless you have a more clever way to only do that check for the first enum

File: sdk/pubkey/src/lib.rs [hunk 11]
```diff
@@ -125,13 +228,14 @@ impl FromStr for Pubkey {
...
+        let mut bytes = [0; PUBKEY_BYTES];
+        let decoded_size = bs58::decode(s)
+            .onto(&mut bytes)
             .map_err(|_| ParsePubkeyError::Invalid)?;
-        if pubkey_vec.len() != mem::size_of::<Pubkey>() {
+        if decoded_size != mem::size_of::<Pubkey>() {
             Err(ParsePubkeyError::WrongSize)
         } else {
-            Pubkey::try_from(pubkey_vec).map_err(|_| ParsePubkeyError::Invalid)
+            Ok(Pubkey(bytes))
```
Comment (joncinque) [inline sdk/pubkey/src/lib.rs:245]: :heart:

File: sdk/pubkey/src/lib.rs [hunk 12]
```diff
@@ -161,13 +266,18 @@ impl TryFrom<Vec<u8>> for Pubkey {
     }
 }
 
+#[cfg(feature = "std")]
```
Comment (joncinque) [inline sdk/pubkey/src/lib.rs]: Is this needed? It should be ok in no-std contexts, just like the `TryFrom<&[u8]>` impl

File: sdk/pubkey/src/lib.rs [hunk 13]
```diff
@@ -639,8 +773,8 @@ impl Pubkey {
             crate::syscalls::sol_log_pubkey(self.as_ref() as *const _ as *const u8)
         };
 
-        #[cfg(not(target_os = "solana"))]
-        crate::program_stubs::sol_log(&self.to_string());
```
Comment (joncinque) [inline sdk/pubkey/src/lib.rs:643]: Side note for me: this is fine for now, but we should restore these stubs at some point, probably through a new crate which just contains the stubs.

**Review by joncinque — COMMENTED** [review]
Looks really good overall! Mostly small things

File: sdk/hash/src/lib.rs (hunk 1 above)
Comment (kevinheavey) [reply sdk/hash/src/lib.rs]: weird, think this found its way in during a rebase or something

File: sdk/pubkey/src/lib.rs (hunk 10 above)
Comment (kevinheavey) [reply sdk/pubkey/src/lib.rs:205]: I used `strum` to iterate through variants and check that strum's `from_repr` matches our `from_i64`:
```rust
for variant in PubkeyError::iter() {
    let variant_i64 = variant as i64;
    assert_eq!(
        PubkeyError::from_repr(variant_i64 as usize),
        PubkeyError::from_i64(variant_i64)
    );
}
```

File: sdk/pubkey/src/lib.rs [hunk 14]
```diff
@@ -1249,4 +1255,26 @@ mod tests {
...
     }
+
+    #[test]
+    fn test_pubkey_error_from_primitive_exhaustive() {
+        for variant in PubkeyError::iter() {
+            let variant_i64 = variant as i64;
+            assert_eq!(
+                PubkeyError::from_repr(variant_i64 as usize),
+                PubkeyError::from_i64(variant_i64)
+            );
```
Comment (joncinque) [inline sdk/pubkey/src/lib.rs:1266]: Let's also have a check for the `From<u64>` implementation
```suggestion
            assert_eq!(
                PubkeyError::from_repr(variant_i64 as usize),
                PubkeyError::from_i64(variant_i64)
            );
            assert_eq!(
                PubkeyError::from(variant_i64 as u64),
                variant
            );
```

**Review by joncinque — COMMENTED** [review]
One last bit, then this is good to go on my side! @yihau , can you accept ownership of `solana-pubkey`?

**Review by joncinque — COMMENTED** [review]
Looks great, thanks! I'll approve once the crate check passes

**Comment by yihau** [comment]
`crate-check` passed!

**Review by joncinque — APPROVED** [review]
Great work, thanks for the contribution!
